Pokémon Red and Blue: Version 6

Alchav opened the pull request:
## What is this fixing or adding?
- Trainersanity has been expanded to include event battles (battles with NPCs that disappear after battle). If your inventory is full and you are unable to receive the item, you can get it later by talking to Oak or having your Pokédex evaluated at a PC.
- Rock Tunnel is no longer forced to be contiguous for Door Shuffle when Randomize Rock Tunnel Layout is enabled.
- In order to more easily program Pokémon location randomization handling, the options to randomize the locations of wild Pokémon, static Pokémon, and starter Pokémon have been condensed to a single option, "Randomize Pokémon Locations." 
- Pokémon placed in static locations are now likely to appear nowhere else.
- Many of what were one-time static Pokémon locations are now repeatable. You can now buy the Route 4 Pokémon Center Pokémon repeatedly, you can perform in-game trades repeatedly, and you can purchase copies of fossils you have turned in at one of the PCs in the Cinnabar Lab Fossil Room. 
- Option to cause EXP to not be split between Pokémon receiving it, instead granting the full amount to each.
- In-game option for alerts for items received from the Multiworld.
- In-game Auto Run option (hold B to walk at normal speed)
- Battle Style (Shift/Set) and Text Speed are set in the yaml and cannot be changed in-game.
- Catch 'Em All option now works when Randomize Pokémon Locations is turned off, making version-exclusives available and making starters and Eevee obtainable repeatedly somewhere.
- Randomize Music option.
- Rods always catch a bite (unless you have no Pokémon!)
- Various vanilla game crashes have been patched.
- Prompt to re-use repel items when they wear off.
- More dialogue has been trimmed from the game.
- Box count is displayed when sending a Pokémon to it in battle.
- Cinnabar Shore glitch has been patched out due to inconsistencies with vanilla game affecting how Missingno. impacts the game, including game crashes.
- The "Key Items Only" option has been removed due to generation failures and difficulty in maintaining code.
- Location Name Groups have been created.
- Item and Location IDs now start at 1.
- Universal Tracker support.

## How was this tested?
Publicly released apworld, private games, using fuzzer to look for clear bugs.

Fixes #4739

Commit:
Dynamic regions for randomized Rock Tunnel

Alchav committed:
Fix Route 3 guard crash

Commit:
Update Randomize Rock Tunnel description

Alchav committed:
Missingo glitch patched

Commit:
Don't allow loop_outs to target each other

Alchav committed:
Merge branch 'pokemon-rb-version-6' of https://github.com/Alchav/Archipelago into pokemon-rb-version-6

Alchav committed:
Trainersanity Event Battles, new options

Alchav committed:
No trapping moves option

Alchav committed:
loop_out_interiors fix

Commit:
bug fixes

Alchav committed:
Split EXP option

Commit:
Rock Tunnel warp fixes

Commit:
Battle Style default is Shift

Commit:
Catch Em All when Pokemon locations not randomized

Alchav committed:
Catch Em All description, etc

Alchav committed:
Merge branch 'pokemon-rb-version-6' of https://github.com/Alchav/Archipelago into pokemon-rb-version-6

Alchav committed:
Condense Pokemon randomization options, improve Catch Em All and add static exclusives

Alchav committed:
Remove Key Items Only

Commit:
Fix Catch Em All

Alchav committed:
Option to Randomize Map Music

Alchav committed:
Randomize Map Music docstring

Commit:
Location name groups

Alchav committed:
Update Pokemon RB doc

Alchav committed:
Debug Options

Alchav committed:
Update Copyright

Alchav committed:
Finish debug option docstring

Alchav committed:
Fix key item check in Indigo Plateau

Alchav committed:
Remove debug text

Alchav committed:
Do not open AP item text box if A button is disabled.

Alchav committed:
Fix Tea crash

Alchav committed:
Fix Rt 4 Pokecenter Mon For Sale text

Alchav committed:
IDs from 1

Alchav committed:
Fix possible softlock

Alchav committed:
Missed Item ID offset

Alchav committed:
Level scaling exception for Dojo gift Pokémon

Exempt-Medic committed:
Merge branch 'main' into pokemon-rb-version-6

Alchav committed:
randomize music crash

Alchav committed:
Merge remote-tracking branch 'origin/pokemon-rb-version-6' into pokemon-rb-version-6

Commit:
Vanilla Music fix

Commit:
Merge remote-tracking branch 'origin/pokemon-rb-version-6' into pokemon-rb-version-6

Commit:
Pokemon Red and Blue Death Link docstring

Alchav committed:
Merge branch 'main' into pokemon-rb-version-6

Commit:
Merge branch 'refs/heads/main' into pokemon-rb-version-6

Commit:
Merge remote-tracking branch 'origin/pokemon-rb-version-6' into pokemon-rb-version-6

Commit:
Fix Player's Mom crash w/ Audio Bank 3 music

Commit:
Simplify PC item placement

Commit:
Merge remote-tracking branch 'origin/pokemon-rb-version-6' into pokemon-rb-version-6

Commit:
Re-fix music randomizer?

Commit:
Pokemon RB: Universal Tracker support

Commit:
Remove outdated part of comment

Commit:
PR up so long it's time to update copyright notice

Commit:
Fix double evolution bug

Commit:
Fix Blackbelt 5 Flag

duckboycool reviewed (approved):
Haven't tested in game, but I don't see any big issues glancing through the code changes quickly, and this branch was stable in #5368 (and fixed some of the current failures).

Mysteryem committed:
Fix mutating global item_groups in get_filler_item_name()

The `banned_items` variable was the `"Unique"`list within the
`item_groups` global. `get_filler_item_name()` could then mutate
`banned_items` through `banned_items.append("Poke Doll")` and
`banned_items += item_groups["Vending Machine Drinks"]`, causing
the contents of the `item_groups` global to be mutated.

This has been fixed by making `banned_items` a shallow copy of
`item_groups["Unique"]`.

Alchav committed:
Merge pull request #10 from Mysteryem/patch-14

Fix mutating global item_groups in get_filler_item_name()

Commit:
Duck's suggestions

Alchav commented on worlds/pokemon_rb/options.py:
```diff
@@ -878,13 +902,45 @@ class RandomizePokemonPalettes(Choice):
+    option_chaos = 3
+
+
+class DebugOptions(OptionList):
```
Done

Alchav commented on worlds/pokemon_rb/encounters.py:
```diff
@@ -2,13 +2,31 @@
+def get_encounter_slots(world, types):
+    encounter_slots = deepcopy([location for location in location_data if location.type in types])
+
+    i = [True, False][world.options.game_version.value]
```
Done

Commit:
Merge branch 'main' into pokemon-rb-version-6

Commit:
Static Repeatable Pokemon not Repeatable Static Pokemon

Commit:
Fix Encounter slot item rule

Commit:
ut_glitch for Dojo gifts

Commit:
Unit tests

palex00 reviewed (changes requested):
I have tried testing Universal Tracker Support for this and it doesn't work.

Using the AP fuzzer's UT-Hook (https://github.com/Eijebong/Archipelago-fuzzer):
python fuzz.py -r50 -g pokemon_rb --hook worlds.tracker.fuzzer_hook:Hook
I tested it 50 times and it was only correct in 7/50 of these. I also tested it with ER disabled (door_shuffle: off) as I thought this might be the issue and this only brought it up to 9/50 successes. I have attached a zip of log failures & YAMLs used for it that can be used to recreate my findings.

Tested on AP 0.6.7 Source with beta apworld as distributed in the RB channel.

[fuzz_output.zip](https://github.com/user-attachments/files/28184112/fuzz_output.zip)

I started looking into this because I'm updating the RB Tracker to work with the Beta and noticed "hey, if it can handle ER in UT, surely I can make it work in RB as well if it doesn't just save the seed to re-gen ER". It does, so I can't, but at least it let me discover that UT doesn't actually work.


I'd love if V6 could also add event flags + Pokémon Tracking to be on par with crystal, firered/leafgreen, platinum, and black & white but I understand if that isn't in scope.
